src/key.rs

```diff
@@ -608,6 +618,247 @@ impl HasByteSource for SecretAsymmetricKey {
     }
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+pub enum SigningKey {
+    SodiumOxideEd25519(SodiumOxideEd25519SecretAsymmetricKey),
+    RingEd25519(RingEd25519SecretAsymmetricKey),
+}
+
+impl StorableType for SigningKey {}
+
+impl From<SigningKey> for Key {
+    fn from(signing_key: SigningKey) -> Self {
+        match signing_key {
+            SigningKey::SodiumOxideEd25519(k) =>
+                Key::Asymmetric(AsymmetricKey::Public(PublicAsymmetricKey::SodiumOxideEd25519(k.public_key().unwrap()))),
+            SigningKey::RingEd25519(k) =>
+                Key::Asymmetric(AsymmetricKey::Public(PublicAsymmetricKey::RingEd25519(k.public_key().unwrap())))
+        }
+    }
+}
+
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum EncryptingKey {
+    SodiumOxideCurve25519(SodiumOxideCurve25519SecretAsymmetricKey),
+    SodiumOxideSymmetricKey(SodiumOxideSymmetricKey),
+}
+
+impl StorableType for EncryptingKey {}
+
+#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
+#[serde(tag = "t", content = "c")]
+pub enum SigningKeyBuilder {
+    SodiumOxideEd25519(SodiumOxideEd25519SecretAsymmetricKeyBuilder),
+    RingEd25519(RingEd25519SecretAsymmetricKeyBuilder),
+}
+
+#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
+#[serde(tag = "t", content = "c")]
+pub enum EncryptingKeyBuilder {
+    SodiumOxideCurve25519(SodiumOxideCurve25519SecretAsymmetricKeyBuilder),
+    SodiumOxideSymmetricKey(SodiumOxideSymmetricKeyBuilder),
+}
+
+impl HasIndex for SigningKey {
+    type Index = Document;
+
+    fn get_index() -> Option<Self::Index> {
```

Comment:
We can use mongo query document to zero in on the exact set of builders this type accepts (https://docs.mongodb.com/manual/tutorial/query-documents/#specify-or-conditions), e.g.

```json
{
  "$or": [
    {
      "c": {
        "builder": {
          "t": "Key",
          "c": {
            "t": "Asymmetric",
            "c": {
              "t": "Secret",
              "c": {
                "t": "SodiumOxideEd25519"
              }
            }
          }
        }
      }
    },
    {
      "c": {
        "builder": {
          "t": "Key",
          "c": {
            "t": "Asymmetric",
            "c": {
              "t": "Secret",
              "c": {
                "t": "RingEd25519"
              }
            }
          }
        }
      }
    }
  ]
}
```

Reply:
I like this less because it's very specific to mongo, and it also needs to be updated every time we want to add a new key type as a SigningKey.